crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,440 @@
+    graph: StableDiGraph<StorageData, ()>,
+    /// Coins -> Transaction that crurrently create the UTXO
+    coins_creators: HashMap<UtxoId, NodeIndex>,
+    /// Contract -> Transaction that currently create the contract
+    contracts_creators: HashMap<ContractId, NodeIndex>,
+}
+
+pub struct GraphConfig {
+    /// The maximum number of transactions per dependency chain
+    pub max_txs_per_chain: u64,
+}
+
+impl GraphStorage {
+    /// Create a new graph storage
+    pub fn new(config: GraphConfig) -> Self {
+        Self {
+            config,
+            graph: StableDiGraph::new(),
+            coins_creators: HashMap::new(),
+            contracts_creators: HashMap::new(),
+        }
+    }
+}
+
+impl GraphStorage {
+    /// Remove a node and all its dependent sub-graph.
+    /// Edit the data of dependencies transactions accordingly.
+    /// Returns the removed transactions.
+    fn remove_node_and_dependent_sub_graph(
+        &mut self,
+        root: NodeIndex,
+    ) -> Result<Vec<PoolTransaction>, Error> {
+        let mut to_traverse = vec![root];
+        let mut to_remove = vec![];
+        while let Some(node_id) = to_traverse.pop() {
+            to_remove.push(node_id);
+            for edge in self
+                .graph
+                .edges_directed(node_id, petgraph::Direction::Outgoing)
+            {
+                to_traverse.push(edge.target());
+            }
+        }
+
+        let mut dependencies: Vec<NodeIndex> = self
+            .graph
+            .neighbors_directed(root, petgraph::Direction::Incoming)
+            .collect();
+        let nb_removed = to_remove.len() as u64;
+        let (gas_removed, tip_removed, removed_transactions) = to_remove.iter().fold(
+            (0, 0, vec![]),
+            |(gas, tip, mut removed_transactions), node_id| {
+                let Some(node) = self.graph.remove_node(*node_id) else {
+                    return (gas, tip, removed_transactions);
+                };
+                removed_transactions.push(node.transaction);
+                (
+                    gas + node.cumulative_gas,
+                    tip + node.cumulative_tip,
+                    removed_transactions,
+                )
+            },
+        );
+
+        while let Some(node_id) = dependencies.pop() {
+            dependencies.extend(
+                self.graph
+                    .neighbors_directed(node_id, petgraph::Direction::Incoming),
+            );
+            let Some(node) = self.graph.node_weight_mut(node_id) else {
+                return Err(Error::Storage(format!(
+                    "Node with id {:?} not found",
+                    node_id
+                )));
+            };
+            node.cumulative_gas = node.cumulative_gas.saturating_sub(gas_removed);
+            node.cumulative_tip = node.cumulative_tip.saturating_sub(tip_removed);
+            node.number_txs_in_chain -= nb_removed;
+        }
+        Ok(removed_transactions)
+    }
```

Comment:
I'd find this function easier to read if we separated the logic for reducing the dependencies cumulative gas, and the logic for removing the dependent sub graph. Something like the following (including the correction of the cumulative gas and tip handling)

```suggestion
    fn remove_node_and_dependent_sub_graph(
        &mut self,
        root_id: NodeIndex,
    ) -> Result<Vec<PoolTransaction>, Error> {
        let Some(root) = self.graph.node_weight(root_id) else {
            return Ok(vec![])
        };

        let gas_removed = root.cumulative_gas;
        let tip_removed = root.cumulative_tip;

        self.reduce_dependencies_cumulative_gas_and_tip(root_id, gas_removed, tip_removed);
        self.remove_dependent_sub_graph(root_id)
    }

    fn reduce_dependencies_cumulative_gas_and_tip(
        &mut self,
        root_id: NodeIndex,
        gas_reduction: u64,
        tip_reduction: u64,
    ) -> Result<(), Error> {
        let Some(root) = self.graph.node_weight_mut(root_id) else {
            return Err(Error::Storage(format!(
                "Node with id {:?} not found",
                root_id
            )));
        };

        root.cumulative_gas.saturating_sub(gas_reduction);
        root.cumulative_tip.saturating_sub(tip_reduction);

        let dependencies: Vec<_> = self
            .graph
            .neighbors_directed(root_id, petgraph::Direction::Incoming)
            .collect();

        for dependency in dependencies {
            self.reduce_dependencies_cumulative_gas_and_tip(dependency, gas_reduction, tip_reduction)?;
        }

        Ok(())
    }

    fn remove_dependent_sub_graph(
        &mut self,
        root_id: NodeIndex,
    ) -> Result<Vec<PoolTransaction>, Error> {
        let dependents: Vec<_> = self
            .graph
            .neighbors_directed(root_id, petgraph::Direction::Outgoing)
            .collect();

        let Some(root) = self.graph.remove_node(root_id) else {
            return Ok(vec![]);
        };

        let mut removed_transactions = vec![root.transaction];

        for dependent in dependents {
            removed_transactions.extend(self.remove_dependent_sub_graph(dependent)?);
        }

        Ok(removed_transactions)
    }
```

The function for reducing the cumulative gas and tip could be generalized and reused for when we add new transactions.

Reply:
Yeah. I agree with your way to split it, Applied.

Reply:
Nice, thank you!